pkg/syncer/volume_health.go

```diff
@@ -77,58 +77,82 @@ func csiGetVolumeHealthStatus(ctx context.Context, k8sclient clientset.Interface
+					log.Errorf("csiGetVolumeHealthStatus: invalid health status %q for volume %q", volHealthStatus, volID)
 				}
-				val, found := pvc.Annotations[annVolumeHealth]
-				_, foundAnnHealthTS := pvc.Annotations[annVolumeHealthTS]
-				if !found || val != volHealthStatus || !foundAnnHealthTS {
-					// VolumeHealth annotation on pvc is changed, set it to new value
-					log.Debugf("csiGetVolumeHealthStatus: update volume health annotation for pvc %s/%s from old value %s to new value %s",
-						pvc.Namespace, pvc.Name, val, volHealthStatus)
-					metav1.SetMetaDataAnnotation(&pvc.ObjectMeta, annVolumeHealth, volHealthStatus)
-					metav1.SetMetaDataAnnotation(&pvc.ObjectMeta, annVolumeHealthTS, time.Now().Format(time.UnixDate))
-					log.Infof("csiGetVolumeHealthStatus: set annotation for health to %s at time %s for pvc %s/%s", volHealthStatus, time.Now().Format(time.UnixDate), pvc.Namespace, pvc.Name)
-					_, err := k8sclient.CoreV1().PersistentVolumeClaims(pvc.Namespace).Update(ctx, pvc, metav1.UpdateOptions{})
+				updateVolumeHealthStatus(ctx, k8sclient, pvc, volHealthStatusAnn)
+			}
+		} else {
+			// Set volume health status as "Inaccessible" when PVC is not found in CNS.
+			// When a Datastore is removed from VC (like vSAN direct disk decommisson with noAction does), the CNS Volumes
+			// on that Datastore are eventually removed from CNS DB, but the PVCs still remain in the K8S cluster.
+			// We are making the design choice of reflecting the CNS cached status of health on the PVC's health annotation
+			// at any given point of time. The vDPp operators are advised to look at the health change timestamp and wait
+			// "long enough" (like an hour) before taking any corrective actions. So if the PVC health is getting updated
+			// every 5 mins, wait for an hour or so before taking any corrective actions. This is an acceptable level
+			// of eventual consistency.
+			updateVolumeHealthStatus(ctx, k8sclient, pvc, common.VolHealthStatusInaccessible)
+		}
+	}
+	log.Infof("GetVolumeHealthStatus: end")
+}
+
+func updateVolumeHealthStatus(ctx context.Context, k8sclient clientset.Interface, pvc *v1.PersistentVolumeClaim, volHealthStatus string) {
+	log := logger.GetLogger(ctx)
+
+	val, found := pvc.Annotations[annVolumeHealth]
+	_, foundAnnHealthTS := pvc.Annotations[annVolumeHealthTS]
+	if !found || val != volHealthStatus || !foundAnnHealthTS {
+		// VolumeHealth annotation on pvc is changed, set it to new value
+		log.Debugf("updateVolumeHealthStatus: update volume health annotation for pvc %s/%s from old value %s to new value %s",
+			pvc.Namespace, pvc.Name, val, volHealthStatus)
+		metav1.SetMetaDataAnnotation(&pvc.ObjectMeta, annVolumeHealth, volHealthStatus)
+		timeNow := time.Now().Format(time.UnixDate)
+		metav1.SetMetaDataAnnotation(&pvc.ObjectMeta, annVolumeHealthTS, timeNow)
+		log.Infof("updateVolumeHealthStatus: set annotation for health to %s at time %s for pvc %s/%s", volHealthStatus, timeNow, pvc.Namespace, pvc.Name)
```

Comment:
Can we move this to debug? Better is to remove it sicne we are anyways adding the health and timestamp as annotation in the PVC.

Reply:
Done